In [58]:
import pandas as pd
import numpy as np
import os
import pickle
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import __version__ as sklearn_version
from sklearn.decomposition import PCA
from sklearn.preprocessing import scale
from sklearn.model_selection import train_test_split, cross_validate, GridSearchCV, learning_curve
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.dummy import DummyRegressor
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.feature_selection import SelectKBest, f_regression
import datetime



Objectives

To investigate the potential causes of road accidents in the US and find patterns to demonstrate the causes. This database will help to reject/accept the following null hypothesis(problem statements)

Null Hypothesis 1: Environmental factors do not contribute to car accidents significantly.

Null Hypothesis 2: During Covid (2020-2021), there has not been any significant change in the pattern of accidents.

Null Hypothesis 3: Environmental factors and accident hotspots are not sufficient to predict real-time accidents.

## Load The Data<a id='3.4_Load_The_Data'></a>

In [59]:
df = pd.read_csv(r'C:\Users\Shweta\Documents\gitrepo\Capstonetwo\USaccidentsdata_cleanednexplored.csv', index_col=False)

In [60]:
# Changing settings in head function to view all columns.
pd.set_option('display.max_columns', 55)
pd.options.display.max_columns = 55
pd.set_option('display.max_columns', None)
df.head(50)

Unnamed: 0.1  Unnamed: 0    ID  Severity           Start_Time  \
0              0           0   A-1         3  2016-02-08 00:37:08   
1              1           1   A-2         2  2016-02-08 05:56:20   
2              2           2   A-3         2  2016-02-08 06:15:39   
3              3           3   A-4         2  2016-02-08 06:51:45   
4              4           4   A-5         3  2016-02-08 07:53:43   
5              5           5   A-6         2  2016-02-08 08:16:57   
6              6           6   A-7         2  2016-02-08 08:15:41   
7              7           7   A-8         2  2016-02-08 11:51:46   
8              8           8   A-9         2  2016-02-08 14:19:57   
9              9           9  A-10         2  2016-02-08 15:16:43   
10            10          10  A-11         2  2016-02-08 15:43:50   
11            11          11  A-12         2  2016-02-08 16:50:57   
12            12          12  A-13         2  2016-02-08 17:27:39   
13            13          13  A-14         2  2016-02-08 17:30:18   
14            14          14  A-15         3  2016-02-08 18:11:11   
15            15          15  A-16         3  2016-02-08 18:11:11   
16            16          16  A-17         3  2016-02-08 19:47:42   
17            17          17  A-18         2  2016-02-08 19:47:42   
18            18          18  A-19         3  2016-02-08 20:13:22   
19            19          19  A-20         4  2016-02-08 21:00:17   
20            20          20  A-21         2  2016-02-08 21:10:10   
21            21          21  A-22         3  2016-02-08 21:30:31   
22            22          22  A-23         2  2016-02-09 05:54:01   
23            23          23  A-24         2  2016-02-09 06:10:59   
24            24          24  A-25         2  2016-02-09 06:46:32   
25            25          25  A-26         2  2016-02-09 07:18:10   
26            26          26  A-27         4  2016-02-09 07:28:15   
27            27          27  A-28         2  2016-02-09 07:25:40   
28            28          28  A-29         3  2016-02-09 07:44:49   
29            29          29  A-30         3  2016-02-09 07:48:40   
30            30          30  A-31         3  2016-02-09 07:48:40   
31            31          31  A-32         2  2016-02-09 08:00:16   
32            32          32  A-33         3  2016-02-09 08:02:55   
33            33          33  A-34         2  2016-02-09 08:18:58   
34            34          34  A-35         2  2016-02-09 09:07:19   
35            35          35  A-36         3  2016-02-09 10:07:49   
36            36          36  A-37         2  2016-02-09 11:09:49   
37            37          37  A-38         2  2016-02-09 11:28:49   
38            38          38  A-39         2  2016-02-09 12:31:08   
39            39          39  A-40         4  2016-02-09 15:33:59   
40            40          40  A-41         2  2016-02-09 17:12:40   
41            41          41  A-42         2  2016-02-09 17:13:58   
42            42          42  A-43         4  2016-02-09 18:20:58   
43            43          43  A-44         4  2016-02-09 18:20:58   
44            44          44  A-45         4  2016-02-09 22:37:48   
45            45          45  A-46         4  2016-02-09 22:37:48   
46            46          46  A-47         4  2016-02-10 06:17:32   
47            47          47  A-48         4  2016-02-10 06:18:49   
48            48          48  A-49         2  2016-02-10 06:29:48   
49            49          49  A-50         2  2016-02-10 08:19:15   

               End_Time  Start_Lat  Start_Lng    End_Lat    End_Lng  \
0   2016-02-08 06:37:08  40.108910 -83.092860  40.112060 -83.031870   
1   2016-02-08 11:56:20  39.865420 -84.062800  39.865010 -84.048730   
2   2016-02-08 12:15:39  39.102660 -84.524680  39.102090 -84.523960   
3   2016-02-08 12:51:45  41.062130 -81.537840  41.062170 -81.535470   
4   2016-02-08 13:53:43  39.172393 -84.492792  39.170476 -84.501798   
5   2016-02-08 14:16:57  39.063240 -84.032430  39.067310 -

In [61]:
# splitting Month&Year columns to Month and Year
df[['Month','Year']] = df["Month&Year"].apply(lambda x: pd.Series(str(x).split(" ")))


In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2776051 entries, 0 to 2776050
Data columns (total 58 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Unnamed: 0.1           int64  
 1   Unnamed: 0             int64  
 2   ID                     object 
 3   Severity               int64  
 4   Start_Time             object 
 5   End_Time               object 
 6   Start_Lat              float64
 7   Start_Lng              float64
 8   End_Lat                float64
 9   End_Lng                float64
 10  Distance(mi)           float64
 11  Description            object 
 12  Number                 float64
 13  Street                 object 
 14  Side                   object 
 15  City                   object 
 16  County                 object 
 17  State                  object 
 18  Zipcode                object 
 19  Country                object 
 20  Timezone               object 
 21  Airport_Code           object 
 22  Weather_Timestamp 

##  Train/Test Split<a id='4.6_Train/Test_Split'></a>

In [172]:
df.shape

(2776051, 58)

In [173]:
len(df) * .7, len(df) * .3

(1943235.7, 832815.2999999999)

In [174]:
X_train, X_test, y_train, y_test = train_test_split(df.drop(columns='Severity'), 
                                                    df.Severity, test_size=0.3, 
                                                    random_state=47)

In [175]:
X_train.shape, X_test.shape

((1943235, 57), (832816, 57))

In [176]:
y_train.shape, y_test.shape

((1943235,), (832816,))

## Imputing missing values of numeric variables with mean 

We can either replace the missing values of Temperature(F) with the mean of the column or still better would be the mean of the zipcode and the month. Instead of just replacing with a mean, we would grouby X_train with zipcodes and Month and then take average temperature. We do the same for other continous numeric variables.

We are missing Wind Chill data for almost 14% of X_train rows and Precipitation for 17% of X_train rows. Deleting these rows could skew the model.

In [177]:
X_train['Temperature(F)'] = X_train['Temperature(F)'].fillna(X_train.groupby(['Month','Zipcode'])['Temperature(F)'].transform('mean'))
X_train['Temperature(F)'] = X_train['Temperature(F)'].fillna(X_train['Temperature(F)'].mean())

In [178]:
X_train['Precipitation(in)'] = X_train['Precipitation(in)'].fillna(X_train.groupby(['Month','Zipcode'])['Precipitation(in)'].transform('mean'))
X_train['Precipitation(in)'] = X_train['Precipitation(in)'].fillna(X_train['Precipitation(in)'].mean())

In [179]:
X_train['Wind_Chill(F)'] = X_train['Wind_Chill(F)'].fillna(X_train.groupby(['Month','Zipcode'])['Wind_Chill(F)'].transform('mean'))
X_train['Wind_Chill(F)'] = X_train['Wind_Chill(F)'].fillna(X_train['Wind_Chill(F)'].mean())

In [180]:
X_train['Humidity(%)'] = X_train['Humidity(%)'].fillna(X_train.groupby(['Month','Zipcode'])['Humidity(%)'].transform('mean'))
X_train['Humidity(%)'] = X_train['Humidity(%)'].fillna(X_train['Humidity(%)'].mean())

In [181]:
X_train['Pressure(in)'] = X_train['Pressure(in)'].fillna(X_train.groupby(['Month','Zipcode'])['Pressure(in)'].transform('mean'))
X_train['Pressure(in)'] = X_train['Pressure(in)'].fillna(X_train['Pressure(in)'].mean())


In [182]:
X_train['Wind_Speed(mph)'] = X_train['Wind_Speed(mph)'].fillna(X_train.groupby(['Month','Zipcode'])['Wind_Speed(mph)'].transform('mean'))
X_train['Wind_Speed(mph)'] = X_train['Wind_Speed(mph)'].fillna(X_train['Wind_Speed(mph)'].mean())

In [183]:
X_train['Visibility(mi)'] = X_train['Visibility(mi)'].fillna(X_train.groupby(['Month','Zipcode'])['Visibility(mi)'].transform('mean'))
X_train['Visibility(mi)'] = X_train['Visibility(mi)'].fillna(X_train['Visibility(mi)'].mean())

In [184]:
X_train.isnull().sum()

Unnamed: 0.1                   0
Unnamed: 0                     0
ID                             0
Start_Time                     0
End_Time                       0
Start_Lat                      0
Start_Lng                      0
End_Lat                        0
End_Lng                        0
Distance(mi)                   0
Description                    0
Number                   1188441
Street                         0
Side                           0
City                          89
County                         0
State                          0
Zipcode                        0
Country                        0
Timezone                       0
Airport_Code                   0
Weather_Timestamp              0
Temperature(F)                 0
Wind_Chill(F)                  0
Humidity(%)                    0
Pressure(in)                   0
Visibility(mi)                 0
Wind_Direction             11926
Wind_Speed(mph)                0
Precipitation(in)              0
Weather_Co

In [185]:
#Check the `dtypes` attribute of `X_train` to verify all features are numeric
X_train.dtypes

Unnamed: 0.1               int64
Unnamed: 0                 int64
ID                        object
Start_Time                object
End_Time                  object
Start_Lat                float64
Start_Lng                float64
End_Lat                  float64
End_Lng                  float64
Distance(mi)             float64
Description               object
Number                   float64
Street                    object
Side                      object
City                      object
County                    object
State                     object
Zipcode                   object
Country                   object
Timezone                  object
Airport_Code              object
Weather_Timestamp         object
Temperature(F)           float64
Wind_Chill(F)            float64
Humidity(%)              float64
Pressure(in)             float64
Visibility(mi)           float64
Wind_Direction            object
Wind_Speed(mph)          float64
Precipitation(in)        float64
Weather_Co

We will need all columns to be numeric. So we begin by deleting the two index columns. And then by converting the non numeric columns to numeric by dummy encoding or One hot encoding. We also going to reduce the numeric values with continous values to bins. 

##  Deleting non-numeric columns not aiding in the study

In [186]:
#Setting column 'ID' as index and deleting the first two index columns for X_train and X_test
X_train.drop(df.columns[[0, 1]], axis = 1, inplace=True)

X_train.set_index('ID', inplace=True)

In [187]:
X_train.head()

Start_Time             End_Time  Start_Lat   Start_Lng  \
ID                                                                           
A-2070655  2020-05-25 20:40:57  2020-05-25 21:10:25  40.703170  -73.932730   
A-2493473  2019-09-12 17:08:00  2019-09-12 18:41:39  34.095787 -117.916578   
A-1698795  2021-02-17 19:10:51  2021-02-17 20:25:54  32.235658 -110.832084   
A-1645872  2021-02-18 18:09:54  2021-02-18 19:45:53  37.099959  -93.232099   
A-1636517  2021-01-14 13:09:00  2021-01-15 00:27:11  25.811613  -80.313412   

             End_Lat     End_Lng  Distance(mi)  \
ID                                               
A-2070655  40.702840  -73.933530         0.048   
A-2493473  34.095787 -117.916578         0.000   
A-1698795  32.235659 -110.832439         0.021   
A-1645872  37.100575  -93.232188         0.043   
A-1636517  25.809433  -80.315273         0.190   

                                                 Description  Number  \
ID                                                                     
A-2070655  Closed at Bushwick Ave - Road closed due to ac...     1.0   
A-2493473                        At E Cypress St - Accident.  4598.0   
A-1698795  Incident on E SPEEDWAY BLVD near N PRUDENCE RD...  7511.0   
A-1645872  Incident on US-65 near E EVANS RD Drive with c...     NaN   
A-1636517  Slow traffic on FL-948 - FL-969 from NW 41st S...  7200.0   

                      Street Side         City       County State     Zipcode  \
ID                                                                              
A-2070655          Bogart St    L     Brooklyn        Kings    NY  11206-3805   
A-2493473   N Lark Ellen Ave    R       Covina  Los Angeles    CA  91722-3133   
A-1698795    E Speedway Blvd    L       Tucson         Pima    AZ  85710-8809   
A-1645872            US-65 N    R  Springfield       Greene    MO       65804   
A-1636517         NW 41st St    L        Miami   Miami-Dade    FL  33166-6712   

          Country     Timezone Airport_Code    Weather_Timestamp  \
ID                                                                 
A-2070655      US   US/Eastern         KJRB  2020-05-25 20:51:00   
A-2493473      US   US/Pacific         KEMT  2019-09-12 16:53:00   
A-1698795      US  US/Mountain         KDMA  2021-02-17 18:58:00   
A-1645872      US   US/Central         KSGF  2021-02-18 17:52:00   
A-1636517      US   US/Eastern         KMIA  2021-01-14 12:53:00   

           Temperature(F)  Wind_Chill(F)  Humidity(%)  Pressure(in)  \
ID                                                                    
A-2070655            63.0      60.140231         81.0         30.09   
A-2493473            88.0      88.000000         37.0         29.78   
A-1698795            55.0      55.000000         17.0         27.07   
A-1645872            18.0       8.000000         71.0         29.01   
A-1636517            65.0      65.000000         68.0         29.99   

           Visibility(mi) Wind_Direction  Wind_Speed(mph)  Precipitation(in)  \
ID                                                                             
A-2070655            10.0            NaN         7.371287                0.0   
A-2493473            10.0            VAR         5.000000                0.0   
A-1698795            10.0              W        18.000000                0.0   
A-1645872            10.0             NW         8.000000                0.0   
A-1636517            10.0            NNW         8.000000                0.0   

          Weather_Condition  Amenity   Bump  Crossing  Give_Way  Junction  \
ID                                                                          
A-2070655              Fair     True  False     False     False     False   
A-2493473              Fair    False  False     False     False     False   
A-1698795              Fair    False  False     False     False     False   
A-1645872              Fair    False  False     False     False     False   
A-1636517            Cloudy    False  False     Fal

In [188]:
X_test.drop(df.columns[[0, 1]], axis = 1, inplace=True)

X_test.set_index('ID', inplace=True)
X_test.head()


Start_Time                       End_Time  \
ID                                                                        
A-1360073  2021-12-12 00:32:00.000000000  2021-12-12 01:50:22.000000000   
A-2048335            2020-11-21 23:27:00            2020-11-22 01:29:05   
A-30665              2016-07-21 11:46:14            2016-07-21 17:46:14   
A-53453              2016-11-10 15:57:37            2016-11-10 21:57:37   
A-2032516            2020-11-24 01:10:20            2020-11-24 02:32:16   

           Start_Lat   Start_Lng    End_Lat     End_Lng  Distance(mi)  \
ID                                                                      
A-1360073  37.522913 -121.946081  37.516056 -121.940620         0.560   
A-2048335  38.870218  -77.046056  38.872018  -77.043376         0.190   
A-30665    29.979460  -95.421990  29.988420  -95.423620         0.627   
A-53453    39.787069  -75.690378  39.785807  -75.686208         0.238   
A-2032516  39.389359  -78.028166  39.405686  -78.023518         1.155   

                                                 Description  Number  \
ID                                                                     
A-1360073  Incident on I-680 SB near SCOTT CRK Drive with...     NaN   
A-2048335  Incident on I-395 NB near MM 10 Left lane bloc...     NaN   
A-30665                     At Airtex Dr/Exit 63 - Accident.     NaN   
A-53453    Closed at Schoolhouse Rd - Road closed due to ...  7250.0   
A-2032516  Incident on I-81 NB near I-81 Drive with caution.     NaN   

                                 Street Side       City      County State  \
ID                                                                          
A-1360073                       I-680 S    R    Fremont     Alameda    CA   
A-2048335  Henry Shirley Memorial Hwy N    R  Arlington   Arlington    VA   
A-30665                          I-45 N    R    Houston      Harris    TX   
A-53453                  Lancaster Pike    R  Hockessin  New Castle    DE   
A-2032516                        I-81 N    R     Inwood    Berkeley    WV   

              Zipcode Country    Timezone Airport_Code    Weather_Timestamp  \
ID                                                                            
A-1360073       94539      US  US/Pacific         KNUQ  2021-12-12 00:35:00   
A-2048335       22202      US  US/Eastern         KDCA  2020-11-21 23:52:00   
A-30665         77090      US  US/Central         KDWH  2016-07-21 11:53:00   
A-53453    19707-9263      US  US/Eastern         KILG  2016-11-10 15:51:00   
A-2032516       25428      US  US/Eastern         KMRB  2020-11-24 00:53:00   

           Temperature(F)  Wind_Chill(F)  Humidity(%)  Pressure(in)  \
ID                                                                    
A-1360073            46.0           41.0         66.0         30.03   
A-2048335            55.0           55.0         62.0         30.39   
A-30665              91.9            NaN         45.0         30.15   
A-53453              57.0            NaN         45.0         29.96   
A-2032516            38.0           38.0         62.0         29.69   

           Visibility(mi) Wind_Direction  Wind_Speed(mph)  Precipitation(in)  \
ID                                                                             
A-1360073            10.0             SE             10.0                0.0   
A-2048335            10.0            NNE              6.0                0.0   
A-30665              10.0           Calm              NaN                NaN   
A-53453              10.0           West              6.9                NaN   
A-2032516            10.0           CALM              0.0                0.0   

          Weather_Condition  Amenity   Bump  Crossing  Give_Way  Junction  \
ID                                                                          
A-1360073            Cloudy    False  False     False     False     False   
A-2048335     Partly Cloudy    False  False     False     False      True   
A-30665       Partly Cloud

In [189]:
#Deleting columns with address and Time which we may not need in predicting the severity.
location_list = ['Distance(mi)', 'Description','Number','Street','Side','City','County','State','Zipcode','Country','Airport_Code']
Time_list =['Start_Time','End_Time','Weather_Timestamp','Weather_Date','Weather_Time','StartTime','EndTime']
X_train.drop(columns=location_list, inplace=True)
X_train.drop(columns=Time_list, inplace=True)

In [190]:
X_train.head()

Start_Lat   Start_Lng    End_Lat     End_Lng     Timezone  \
ID                                                                     
A-2070655  40.703170  -73.932730  40.702840  -73.933530   US/Eastern   
A-2493473  34.095787 -117.916578  34.095787 -117.916578   US/Pacific   
A-1698795  32.235658 -110.832084  32.235659 -110.832439  US/Mountain   
A-1645872  37.099959  -93.232099  37.100575  -93.232188   US/Central   
A-1636517  25.811613  -80.313412  25.809433  -80.315273   US/Eastern   

           Temperature(F)  Wind_Chill(F)  Humidity(%)  Pressure(in)  \
ID                                                                    
A-2070655            63.0      60.140231         81.0         30.09   
A-2493473            88.0      88.000000         37.0         29.78   
A-1698795            55.0      55.000000         17.0         27.07   
A-1645872            18.0       8.000000         71.0         29.01   
A-1636517            65.0      65.000000         68.0         29.99   

           Visibility(mi) Wind_Direction  Wind_Speed(mph)  Precipitation(in)  \
ID                                                                             
A-2070655            10.0            NaN         7.371287                0.0   
A-2493473            10.0            VAR         5.000000                0.0   
A-1698795            10.0              W        18.000000                0.0   
A-1645872            10.0             NW         8.000000                0.0   
A-1636517            10.0            NNW         8.000000                0.0   

          Weather_Condition  Amenity   Bump  Crossing  Give_Way  Junction  \
ID                                                                          
A-2070655              Fair     True  False     False     False     False   
A-2493473              Fair    False  False     False     False     False   
A-1698795              Fair    False  False     False     False     False   
A-1645872              Fair    False  False     False     False     False   
A-1636517            Cloudy    False  False     False     False     False   

           No_Exit  Railway  Roundabout  Station   Stop  Traffic_Calming  \
ID                                                                         
A-2070655    False    False       False    False  False            False   
A-2493473    False    False       False    False  False            False   
A-1698795    False    False       False    False  False            False   
A-1645872    False    False       False    False  False            False   
A-1636517    False    False       False    False  False            False   

           Traffic_Signal  Turning_Loop Sunrise_Sunset Civil_Twilight  \
ID                                                                      
A-2070655            True         False          Night            Day   
A-2493473           False         False            Day            Day   
A-1698795           False         False          Night          Night   
A-1645872           False         False          Night            Day   
A-1636517           False         False            Day            Day   

          Nautical_Twilight Astronomical_Twilight Month&Year   StartDate  \
ID                                                                         
A-2070655               Day                   Day   May 2020  2020-05-25   
A-2493473               Day                   Day   Sep 2019  2019-09-12   
A-1698795             Night                   Day   Feb 2021  2021-02-17   
A-1645872               Day                   Day   Feb 2021  2021-02-18   
A-1636517               Day                   Day   Jan 2021  2021-01-14   

              EndDate Month  Year  
ID                                 
A-2070655  2020-05-25   May  2020  
A-2493473  2019-09-12   Sep  2019  
A-1698795  2021-02-17   Feb  2021  
A-1645872  2021-02-18   Feb  2021  
A-1636517  2021-01-15   Jan  2021

In [191]:
X_train.shape

(1943235, 36)

In [192]:
#Deleting columns with address and Time in X_Test which we may not need in predicting the severity.

X_test.drop(columns=location_list, inplace=True)
X_test.drop(columns=Time_list, inplace=True)

In [193]:
X_test.shape # Are we doing the right thing by deleting the columns in the X_test as well? Mentor

(832816, 36)

##  Limiting Categorical Variables

In [194]:
X_train['Wind_Direction'].unique()

array([nan, 'VAR', 'W', 'NW', 'NNW', 'SW', 'East', 'S', 'CALM', 'NNE',
       'SE', 'WSW', 'E', 'WNW', 'N', 'South', 'ESE', 'NE', 'North', 'SSE',
       'Calm', 'ENE', 'Variable', 'SSW', 'West'], dtype=object)

Some values have repeated and can be corrected. For instance, S and South mean the same. So we can replace both by S. Similarly, VAR and Variable mean the same. nan can be written as NA.

In [195]:
X_train['Wind_Direction'].replace('VAR','Variable', inplace=True)
X_train['Wind_Direction'].replace('East','E', inplace=True)
X_train['Wind_Direction'].replace('South','S', inplace=True)
X_train['Wind_Direction'].replace('North','N', inplace=True)
X_train['Wind_Direction'].replace('West','W', inplace=True)
X_train['Wind_Direction'].replace('CALM','Calm', inplace=True)
X_train.Wind_Direction.fillna('NA', inplace=True)

In [196]:
X_train['Wind_Direction'].unique()

array(['NA', 'Variable', 'W', 'NW', 'NNW', 'SW', 'E', 'S', 'Calm', 'NNE',
       'SE', 'WSW', 'WNW', 'N', 'ESE', 'NE', 'SSE', 'ENE', 'SSW'],
      dtype=object)

As we recall from the exploratory data analysis that most accidents were recorded on days when the weather condition was 'Fair' and 'Clear', we can start binning the values in the column Weather_Condition.

In [197]:
X_train.Weather_Condition.value_counts()

Fair                              771665
Mostly Cloudy                     254102
Cloudy                            242984
Partly Cloudy                     174317
Clear                             121088
                                   ...  
Sand / Dust Whirlwinds / Windy         1
Heavy Thunderstorms and Snow           1
Heavy Rain Shower / Windy              1
Duststorm                              1
Heavy Blowing Snow                     1
Name: Weather_Condition, Length: 121, dtype: int64

In [198]:
X_train.Weather_Condition.value_counts(normalize=True).mul(100).round(2).astype(str) + '%'

Fair                              39.89%
Mostly Cloudy                     13.14%
Cloudy                            12.56%
Partly Cloudy                      9.01%
Clear                              6.26%
                                   ...  
Sand / Dust Whirlwinds / Windy      0.0%
Heavy Thunderstorms and Snow        0.0%
Heavy Rain Shower / Windy           0.0%
Duststorm                           0.0%
Heavy Blowing Snow                  0.0%
Name: Weather_Condition, Length: 121, dtype: object

We will set a threshold_percent of 2500 and create a new column 'Weather_Condition_new' which will have all variables with frequency >2500. The ones less than 2500 will be merged to a new variable 'Other'. Limit 0.1% corresponds to variable 'Light Snow / Windy' with a count of 2136(refer EDA).

In [199]:
# Setting all the null values with NA
X_train.Weather_Condition.fillna('NA', inplace=True)

In [200]:
# set up the threshold percent
threshold_percent = 0.1

series = pd.value_counts(X_train['Weather_Condition'])
mask = (series / series.sum() * 100).lt(threshold_percent)
X_train = X_train.assign(Weather_Condition_new = np.where(X_train['Weather_Condition'].isin(series[mask].index),'Other', X_train['Weather_Condition']))

In [201]:
X_train.Weather_Condition_new.value_counts()

Fair                       771665
Mostly Cloudy              254102
Cloudy                     242984
Partly Cloudy              174317
Clear                      121088
Light Rain                  89574
Overcast                    59133
Scattered Clouds            31407
Light Snow                  30642
Fog                         28830
Haze                        25371
Rain                        21540
Other                       14755
Fair / Windy                10486
NA                           8856
Heavy Rain                   8277
Smoke                        5076
Light Drizzle                4951
Thunder in the Vicinity      4771
Cloudy / Windy               4675
T-Storm                      4560
Mostly Cloudy / Windy        4409
Thunder                      4164
Snow                         3703
Light Rain with Thunder      3642
Wintry Mix                   2703
Partly Cloudy / Windy        2682
Heavy T-Storm                2479
Light Rain / Windy           2393
Name: Weather_

Now we have a new column in X_train with 'Weather_Condition_new' with a new varaible 'Other' with 14755 value counts. 

In [202]:
#Note: This could affect X_test. Should we add a new column there as well and start limiting variables in Wind Direction and Weather Condition columns OR should we just rename the Weather_Condition_new in X_train to Weather_Condition? Mentor Discussion.

In [203]:
# could the percentages vary in X_test and create a different categories?

In [204]:
X_train.isnull().sum()

Start_Lat                   0
Start_Lng                   0
End_Lat                     0
End_Lng                     0
Timezone                    0
Temperature(F)              0
Wind_Chill(F)               0
Humidity(%)                 0
Pressure(in)                0
Visibility(mi)              0
Wind_Direction              0
Wind_Speed(mph)             0
Precipitation(in)           0
Weather_Condition           0
Amenity                     0
Bump                        0
Crossing                    0
Give_Way                    0
Junction                    0
No_Exit                     0
Railway                     0
Roundabout                  0
Station                     0
Stop                        0
Traffic_Calming             0
Traffic_Signal              0
Turning_Loop                0
Sunrise_Sunset           1595
Civil_Twilight           1595
Nautical_Twilight        1595
Astronomical_Twilight    1595
Month&Year                  0
StartDate                   0
EndDate   

In [205]:
#changing the boolean variables to integer
X_train.replace(False, 0, inplace=True)
X_train.replace(True, 1, inplace=True)

In [206]:
X_train.head()

Start_Lat   Start_Lng    End_Lat     End_Lng     Timezone  \
ID                                                                     
A-2070655  40.703170  -73.932730  40.702840  -73.933530   US/Eastern   
A-2493473  34.095787 -117.916578  34.095787 -117.916578   US/Pacific   
A-1698795  32.235658 -110.832084  32.235659 -110.832439  US/Mountain   
A-1645872  37.099959  -93.232099  37.100575  -93.232188   US/Central   
A-1636517  25.811613  -80.313412  25.809433  -80.315273   US/Eastern   

           Temperature(F)  Wind_Chill(F)  Humidity(%)  Pressure(in)  \
ID                                                                    
A-2070655            63.0      60.140231         81.0         30.09   
A-2493473            88.0      88.000000         37.0         29.78   
A-1698795            55.0      55.000000         17.0         27.07   
A-1645872            18.0       8.000000         71.0         29.01   
A-1636517            65.0      65.000000         68.0         29.99   

           Visibility(mi) Wind_Direction  Wind_Speed(mph)  Precipitation(in)  \
ID                                                                             
A-2070655            10.0             NA         7.371287                0.0   
A-2493473            10.0       Variable         5.000000                0.0   
A-1698795            10.0              W        18.000000                0.0   
A-1645872            10.0             NW         8.000000                0.0   
A-1636517            10.0            NNW         8.000000                0.0   

          Weather_Condition  Amenity  Bump  Crossing  Give_Way  Junction  \
ID                                                                         
A-2070655              Fair        1     0         0         0         0   
A-2493473              Fair        0     0         0         0         0   
A-1698795              Fair        0     0         0         0         0   
A-1645872              Fair        0     0         0         0         0   
A-1636517            Cloudy        0     0         0         0         0   

           No_Exit  Railway  Roundabout  Station  Stop  Traffic_Calming  \
ID                                                                        
A-2070655        0        0           0        0     0                0   
A-2493473        0        0           0        0     0                0   
A-1698795        0        0           0        0     0                0   
A-1645872        0        0           0        0     0                0   
A-1636517        0        0           0        0     0                0   

           Traffic_Signal  Turning_Loop Sunrise_Sunset Civil_Twilight  \
ID                                                                      
A-2070655               1             0          Night            Day   
A-2493473               0             0            Day            Day   
A-1698795               0             0          Night          Night   
A-1645872               0             0          Night            Day   
A-1636517               0             0            Day            Day   

          Nautical_Twilight Astronomical_Twilight Month&Year   StartDate  \
ID                                                                         
A-2070655               Day                   Day   May 2020  2020-05-25   
A-2493473               Day                   Day   Sep 2019  2019-09-12   
A-1698795             Night                   Day   Feb 2021  2021-02-17   
A-1645872               Day                   Day   Feb 2021  2021-02-18   
A-1636517               Day                   Day   Jan 2021  2021-01-14   

              EndDate Month  Year Weather_Condition_new  
ID                                                       
A-2070655  2020-05-25   May  2020                  Fair  
A-2493473  2019-09-12   Sep  2019                  Fair  
A-1698795  2021-02-17   Feb  2021                  Fair  
A-1645872  2021-02-18   Feb  2021                  Fa

Note as in the case of boolean columns, we can also convert the columns with Day and Night entries to 0 and 1. But we prefer to leave them intact.

In [207]:
X_train["Sunrise_Sunset"].fillna("NA", inplace = True)

In [208]:
X_train.head()

Start_Lat   Start_Lng    End_Lat     End_Lng     Timezone  \
ID                                                                     
A-2070655  40.703170  -73.932730  40.702840  -73.933530   US/Eastern   
A-2493473  34.095787 -117.916578  34.095787 -117.916578   US/Pacific   
A-1698795  32.235658 -110.832084  32.235659 -110.832439  US/Mountain   
A-1645872  37.099959  -93.232099  37.100575  -93.232188   US/Central   
A-1636517  25.811613  -80.313412  25.809433  -80.315273   US/Eastern   

           Temperature(F)  Wind_Chill(F)  Humidity(%)  Pressure(in)  \
ID                                                                    
A-2070655            63.0      60.140231         81.0         30.09   
A-2493473            88.0      88.000000         37.0         29.78   
A-1698795            55.0      55.000000         17.0         27.07   
A-1645872            18.0       8.000000         71.0         29.01   
A-1636517            65.0      65.000000         68.0         29.99   

           Visibility(mi) Wind_Direction  Wind_Speed(mph)  Precipitation(in)  \
ID                                                                             
A-2070655            10.0             NA         7.371287                0.0   
A-2493473            10.0       Variable         5.000000                0.0   
A-1698795            10.0              W        18.000000                0.0   
A-1645872            10.0             NW         8.000000                0.0   
A-1636517            10.0            NNW         8.000000                0.0   

          Weather_Condition  Amenity  Bump  Crossing  Give_Way  Junction  \
ID                                                                         
A-2070655              Fair        1     0         0         0         0   
A-2493473              Fair        0     0         0         0         0   
A-1698795              Fair        0     0         0         0         0   
A-1645872              Fair        0     0         0         0         0   
A-1636517            Cloudy        0     0         0         0         0   

           No_Exit  Railway  Roundabout  Station  Stop  Traffic_Calming  \
ID                                                                        
A-2070655        0        0           0        0     0                0   
A-2493473        0        0           0        0     0                0   
A-1698795        0        0           0        0     0                0   
A-1645872        0        0           0        0     0                0   
A-1636517        0        0           0        0     0                0   

           Traffic_Signal  Turning_Loop Sunrise_Sunset Civil_Twilight  \
ID                                                                      
A-2070655               1             0          Night            Day   
A-2493473               0             0            Day            Day   
A-1698795               0             0          Night          Night   
A-1645872               0             0          Night            Day   
A-1636517               0             0            Day            Day   

          Nautical_Twilight Astronomical_Twilight Month&Year   StartDate  \
ID                                                                         
A-2070655               Day                   Day   May 2020  2020-05-25   
A-2493473               Day                   Day   Sep 2019  2019-09-12   
A-1698795             Night                   Day   Feb 2021  2021-02-17   
A-1645872               Day                   Day   Feb 2021  2021-02-18   
A-1636517               Day                   Day   Jan 2021  2021-01-14   

              EndDate Month  Year Weather_Condition_new  
ID                                                       
A-2070655  2020-05-25   May  2020                  Fair  
A-2493473  2019-09-12   Sep  2019                  Fair  
A-1698795  2021-02-17   Feb  2021                  Fair  
A-1645872  2021-02-18   Feb  2021                  Fa

In [209]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1943235 entries, A-2070655 to A-908872
Data columns (total 37 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Start_Lat              float64
 1   Start_Lng              float64
 2   End_Lat                float64
 3   End_Lng                float64
 4   Timezone               object 
 5   Temperature(F)         float64
 6   Wind_Chill(F)          float64
 7   Humidity(%)            float64
 8   Pressure(in)           float64
 9   Visibility(mi)         float64
 10  Wind_Direction         object 
 11  Wind_Speed(mph)        float64
 12  Precipitation(in)      float64
 13  Weather_Condition      object 
 14  Amenity                int64  
 15  Bump                   int64  
 16  Crossing               int64  
 17  Give_Way               int64  
 18  Junction               int64  
 19  No_Exit                int64  
 20  Railway                int64  
 21  Roundabout             int64  
 22  Station       

In [210]:
X_train.isnull().sum()

Start_Lat                   0
Start_Lng                   0
End_Lat                     0
End_Lng                     0
Timezone                    0
Temperature(F)              0
Wind_Chill(F)               0
Humidity(%)                 0
Pressure(in)                0
Visibility(mi)              0
Wind_Direction              0
Wind_Speed(mph)             0
Precipitation(in)           0
Weather_Condition           0
Amenity                     0
Bump                        0
Crossing                    0
Give_Way                    0
Junction                    0
No_Exit                     0
Railway                     0
Roundabout                  0
Station                     0
Stop                        0
Traffic_Calming             0
Traffic_Signal              0
Turning_Loop                0
Sunrise_Sunset              0
Civil_Twilight           1595
Nautical_Twilight        1595
Astronomical_Twilight    1595
Month&Year                  0
StartDate                   0
EndDate   

In [211]:
# subsetting X_train to only contain numeric relevant columns.
X_train=X_train[['Temperature(F)', 'Wind_Chill(F)', 'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction', 'Wind_Speed(mph)', 'Precipitation(in)', 'Amenity', 'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway', 'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal', 'Turning_Loop', 'Sunrise_Sunset', 'Weather_Condition_new']]

We have dropped the following
Data about Latitude and Longitude - Columns Start_Lat, End_Lat & Start_Lng, End_Lng.# ask mentor
Timezone- Timezone has no relevance in predicting an accident
Twilight Columns- Civil_Twilight, Nautical_Twilight and Astronomical_Twilight. These columns are currently not aiding in the study.
Month and Year columns - Month&Year, Weather Timestamp, Start_time, End_Time- Again time is not a factor in the study.

For 8856 missing values of Weather Condition, we will replace them with NA and for the 1595 missing values of Surise_Sunset, we will set it as NA and apply dummy encoding to it.
Since the boolean variables have already been converted to 1's or 0's, we will apply dummy encoding to Wind_Direction, Weather Condition and Sunrise_Sunset columns.

In [212]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1943235 entries, A-2070655 to A-908872
Data columns (total 23 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Temperature(F)         float64
 1   Wind_Chill(F)          float64
 2   Humidity(%)            float64
 3   Pressure(in)           float64
 4   Visibility(mi)         float64
 5   Wind_Direction         object 
 6   Wind_Speed(mph)        float64
 7   Precipitation(in)      float64
 8   Amenity                int64  
 9   Bump                   int64  
 10  Crossing               int64  
 11  Give_Way               int64  
 12  Junction               int64  
 13  No_Exit                int64  
 14  Railway                int64  
 15  Roundabout             int64  
 16  Station                int64  
 17  Stop                   int64  
 18  Traffic_Calming        int64  
 19  Traffic_Signal         int64  
 20  Turning_Loop           int64  
 21  Sunrise_Sunset         object 
 22  Weather_Condit

In [214]:
X_train.isnull().sum()

Temperature(F)           0
Wind_Chill(F)            0
Humidity(%)              0
Pressure(in)             0
Visibility(mi)           0
Wind_Direction           0
Wind_Speed(mph)          0
Precipitation(in)        0
Amenity                  0
Bump                     0
Crossing                 0
Give_Way                 0
Junction                 0
No_Exit                  0
Railway                  0
Roundabout               0
Station                  0
Stop                     0
Traffic_Calming          0
Traffic_Signal           0
Turning_Loop             0
Sunrise_Sunset           0
Weather_Condition_new    0
dtype: int64

## Dummy Encoding

Since there are no missing values in X_train categorical varaibles, we will apply dummy encoding to the two categorical columns 'Wind_Speed' and 'Weather_Condition_new'

In [215]:
X_train=pd.get_dummies(X_train, columns=['Wind_Direction'], drop_first = True, prefix = 'WD')
X_train=pd.get_dummies(X_train, columns=['Weather_Condition_new'], drop_first = True, prefix = 'WC')
X_train=pd.get_dummies(X_train, columns=['Sunrise_Sunset'], drop_first = True, prefix = 'SS')

In [216]:
X_train.isnull().sum()

Temperature(F)                0
Wind_Chill(F)                 0
Humidity(%)                   0
Pressure(in)                  0
Visibility(mi)                0
                             ..
WC_Thunder                    0
WC_Thunder in the Vicinity    0
WC_Wintry Mix                 0
SS_NA                         0
SS_Night                      0
Length: 68, dtype: int64

In [ ]:
X_train.head()

In [217]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1943235 entries, A-2070655 to A-908872
Data columns (total 68 columns):
 #   Column                      Dtype  
---  ------                      -----  
 0   Temperature(F)              float64
 1   Wind_Chill(F)               float64
 2   Humidity(%)                 float64
 3   Pressure(in)                float64
 4   Visibility(mi)              float64
 5   Wind_Speed(mph)             float64
 6   Precipitation(in)           float64
 7   Amenity                     int64  
 8   Bump                        int64  
 9   Crossing                    int64  
 10  Give_Way                    int64  
 11  Junction                    int64  
 12  No_Exit                     int64  
 13  Railway                     int64  
 14  Roundabout                  int64  
 15  Station                     int64  
 16  Stop                        int64  
 17  Traffic_Calming             int64  
 18  Traffic_Signal              int64  
 19  Turning_Loop     

In [ ]:
Now the data is ready for modelling. We can apply Min Max Scaler to the continuous variables. 

In [ ]:
# save the data to a new csv file
datapath = 'C:\\Users\\Shweta\\Documents\\gitrepo\\Capstonetwo\\'
df.to_csv(datapath+'USaccidentsdata_preprocessed.csv')